job-manager / job-info: get job data changes from job-manager to job-info so job listings will have up to date information

chu11 opened the pull request:
Per discussion in #3052, the `job-manager` did not have a way to inform the `job-info` module that a job's priority had changed, thus leading to invalid job listings.

Several issues were discussed in #3052, this PR just tried to look at the notification of the priority change between `job-manager` & `job-info`.

For prototyping purposes, I implemented about the most straightforward thing I could think of, adding a new `job-updates` event that the `job-manager` publishes.  It publishes a change record of "jobid, timestamp, field that changed, value", such as "JOBID, TIMESTAMP, "priority", 14" would be a potential change.  I transfer the "value" as a json object.  In the future, I believe a json null can serve as "cache invalidate", when the data that has changed isn't known to the `job-manager`.

In the `job-info` module, it has to keep track of the timestamp of when the `priority` field was last set, to avoid any racing that can occur.

- b/c the priority field is (currently) the only job data that can change, I think this implementation is ok.  But as more data can change, this approach will become cumbersome, managing timestamps of when all of the job data was last updated.  So a refactoring would be necessary.

- The `job-updates` and `job-annotations` events could be combined into one event.  If we go with this approach, I think that would be a commit to add to the top of this series.

Any thoughts / comments?

garlick commented:
Just read through #3052 again.

> The job-updates and job-annotations events could be combined into one event. If we go with this approach, I think that would be a commit to add to the top of this series.

That makes a lot of sense.

The one thought that occurs to me regarding a combined protocol for "job updates" is that we did invest some design effort already in job eventlogs.  Maybe rather than creating a new update protocol, we should just publish eventlog entries (adding jobid to context)?

For that matter, would it be better to just "cc" all job eventlog updates to an in-memory circular buffer in the job manager that job-info could consume via a streaming RPC?  Call it the job manager's "journal".  Published events are easy to work with but the big downside is that they are broadcast across the instance, for naught if there is only the one consumer.

For the annotations and upcoming fair share priority changes that probably would not be appropriate to land in the job's eventlog, we still add an eventlog entry to the in-memory journal?

chu11 commented:
> The one thought that occurs to me regarding a combined protocol for "job updates" is that we did invest some design effort already in job eventlogs. Maybe rather than creating a new update protocol, we should just publish eventlog entries (adding jobid to context)?

I like this idea.  The eventlog entry is most of what is needed anyways for the update, so it makes sense.

> For that matter, would it be better to just "cc" all job eventlog updates to an in-memory circular buffer in the job manager that job-info could consume via a streaming RPC? Call it the job manager's "journal". Published events are easy to work with but the big downside is that they are broadcast across the instance, for naught if there is only the one consumer.

At one point in time I did something similar to this for the annotations.  IIRC it worked, but I disliked it b/c unlike most streaming RPCs this streaming RPC had to "last forever".  i.e. gotta always re-connect to `job-manager` if anything goes wrong.  Was a bit annoying to deal with, but now that I think of it, I may have overthought my implementation.  Probably just having a timer reactor that will just connect to the `job-manager` if it's not up is more than enough.

> For the annotations and upcoming fair share priority changes that probably would not be appropriate to land in the job's eventlog, we still add an eventlog entry to the in-memory journal?

Sounds like a decent idea.

garlick commented:
OK, let me open an issue on that idea.

chu11 commented:
re-pushed with a rough implementation (needs to be cleaned up) where the job-info module gets information about job priority changes via a new `job-manager.updates` streaming RPC.  The "record" of changes is done via an eventlog entry per discussion in #3209.

Currently the streaming RPC sends out one eventlog entry every time there is a priority change.  This can be bad / slow b/c there can be a resulting re-sort of the pending job queue in the job-info module everytime it receives a priority change.

I was going to update the job-manager to "batch"  send priority changes, not so dissimilar from the job-manager's current job state transistions batching.  BUT then I realized, this same problem exists in the job-manager.  Changing a job's priority is a single RPC (`job-manager.priority` rpc) and that can result in a re-sorting of the job-manager's alloc queue each time a priority is changed.

As discussed in #3052 there are multiple issues in play.  How to update many priorities in the job-manager at once (RPC wise and internal data structure wise), how to send that data to the job-info manager (both to be done + done efficiently), and how to re-sort in the job-info module (data structure wise).

Many of these solutions are still TBD.  So I think that the current non-optimal implementation may be a good round 1 implementation to atleast solve the initial problem brought up in #3052.  But we should split out new issues for the performance issue on multiple fronts.  I suspect how to update a large number of job priorities at one time in the job-manager will lead us to how to update large number of job priorities in job-info as well.

garlick commented:
If the priority of one job changes, the reordering of the alloc queue will not change hte priorities of other jobs, so that's still one event, right?

chu11 commented:
@garlick correct.  Sorry, perhaps I didn't explain my thoughts correctly.  Basically the issue I was thinking about was, this is how to change the priority of multiple jobs right now (obviously in command line form, but underneath its just the RPCs):

```
for id in 1 2 3 4 5
do
    flux job priority $id 31
done
```

this will lead to 5 re-sorts of the alloc queue.  Which is probably not good.

And in my PR the above will lead to 5 events sent to the job-info module, leading to 5 re-sorts of the job listing pending queue.  Also probably not good.

garlick commented:
The list is stored sorted, so it's just a remove and insert, lt least in the job manager case.

We might be generating many eventlog entries in this "journal" thing if we change a lot of priorities at once, e.g. via bank scheduling, but I'm not super worried about say thousands of short messages sent point to point on the same node.   Maybe I'm not envisioning the full horror though.  (Like with a million jobs in the queue, and updating the banks ever 5 minutes?)

As you say batching can help...

garlick commented:
p.s. I was picturing that we would just add code to send all events that go to job eventlogs to this journal, and then we have a generic method of tracking all changes to all jobs, rather than a one off for priority.  Maybe that's where you're going but wouldn't it be easier to just attach to the current batch mechanism and send out all events now?

chu11 commented:
@garlick ohhh, you call `zlistx_reorder ()` instead of `zlistx_sort ()`, which is much better.  I'll do that as well.

chu11 commented:
> p.s. I was picturing that we would just add code to send all events that go to job eventlogs to this journal, and then we have a generic method of tracking all changes to all jobs, rather than a one off for priority. Maybe that's where you're going but wouldn't it be easier to just attach to the current batch mechanism and send out all events now?

hmmm, I didn't think about that, b/c that's a lot of events given the majority aren't needed by job-info.  Perhaps if there was a filtering mechanism, so that only certain eventlog entries are sent out on the "updates" requests?

garlick commented:
Great idea.  A filter option in the request (like allow/deny list) makes a lot of sense!

chu11 commented:
Just re-pushed.  This version renames the `job-manager.updates` RPC to `job-manager.events`.  `job-manager.events` is now also tied to the `job-manager` batch infrastructure, so after the timer goes off is when an array of events (instead of just 1 event before) are sent for each listener.  Each listener can input a `filters` option to limit which events they only want to listen to.

I think this infrastructure works out really well with the circular buffer in #3209 being considered for the future.

- Because each listener has its own filters, each "listener" has their own `json events` array containing the events it will send on the next batch timer send.  This is in contrast to keeping everything in one list / buffer and parsing / filtering everything when its time to send.

- B/c of the above, the "circular buffer" doesn't have to be so complex and a bunch of "where was I in the buffer" pointer management has to be done.  A simple zlist or json array can service as the "buffer" of the most recent N events.  Will try and prototype this afternoon.

- As a follow up to this PR, I think we can merge annotations into this infrastructure and eliminate the `job-annotations` event.

garlick commented:
> each "listener" has their own json events array containing the events it will send on the next batch timer send.

That sounds great.  I was picturing using the actua newline separated, RFC-definedl eventlog format in basically a cbuf, but I'm on board with this approach if it works out better.  Hopefully the events in the array are objects not string-encoded JSON?  (Double encoding JSON should probably be an official anti-pattern by now :-)

chu11 commented:
> That sounds great. I was picturing using the actual newline separated, RFC-defined eventlog format in basically a cbuf, but I'm on board with this approach if it works out better. Hopefully the events in the array are objects not string-encoded JSON? (Double encoding JSON should probably be an official anti-pattern by now :-)

Yup, arrays of objects, and super cheap to just `json_incref()` the objects into the arrays.

The one inefficiency is the regularly looping over the "filters" for comparison.

```
            json_array_foreach (el->filters, index, value) {
                const char *str = json_string_value (value);
                if (!strcmp (str, name)) {
                    if (json_array_append (el->events, entry) < 0)
                        goto nomem;
                    event->batch->events_to_send = true;
                }
            }
```

I could do some hashing tricks in the future to eliminate the regular looping, but figure this was fine for now.

chu11 commented:
Just re-pushed, doing some cleanup + fixes + adding more tests.  I removed WIP b/c we're approaching something that I think will be our final solution.

chu11 commented:
As an side, am working on an additional test to cover the job-info module reconnecting to the job-manager in the event something goes wrong (e.g. job-manager re-loads), but hitting some issues (first of which was #3218)

chu11 commented:
re-pushed adding 1 more test for "reconnection" of the job-info module to the job-manager being reloaded.  Perhaps this additional test was overkill.

garlick commented:
LMK when you're ready for a review @chu11.

chu11 commented:
@garlick i think its ready for a review, thanks

garlick commented on src/modules/job-manager/event.c:
```diff
@@ -548,6 +601,47 @@ static int get_timestamp_now (double *timestamp)
+            size_t index;
+            json_t *value;
+
+            json_array_foreach (el->filters, index, value) {
+                const char *str = json_string_value (value);
+                if (!strcmp (str, name)) {
+                    if (json_array_append (el->events, entry) < 0)
+                        goto nomem;
+                    event->batch->events_to_send = true;
+                }
+            }
```
Perhaps if instead of a filters array, the request included optional allow or deny _objects_, then jansson's hashing could be utilized instead of walking the array for each element?  E.g.
```json
{
  "allow":{
    "priority":1,
    "foo":1
   }
}
```
The test then just becomes
```c
if ((el->allow && json_object_get (el->allow, name)) 
        || (el->deny && !json_object_get (el->deny, name)))
    json_array_append (el->events, entry)...
```

garlick replied:
Er that logic is slightly wrong since it doesn't cover if both allow and deny are missing, but my point wasn't to suggest allow/deny semantics - I _think_ that's a well undestood design pattern?

garlick commented on src/modules/job-manager/job-manager.c:
```diff
@@ -61,6 +61,24 @@ void disconnect_rpc (flux_t *h,
+    struct job_manager *ctx = arg;
+    int listeners = event_listeners_count (ctx->event);
+    if (flux_respond_pack (h, msg, "{s:i}",
+                           "listeners", listeners) < 0) {
+        flux_log_error (h, "%s: flux_respond_pack", __FUNCTION__);
```
Maybe change the key to "event-listeners" in case we add another streaming RPC to job manager that we want to count here.

garlick commented on src/modules/job-manager/priority.c:
```diff
@@ -97,7 +97,8 @@ void priority_handle_request (flux_t *h,
+                             "id", job->id,
                              "userid", cred.userid,
                              "priority", priority) < 0)
         goto error;
```
We're going to need every event in the events RPC stream to carry a job ID.

If we change every `event_job_post_pack()` call, then we'll be including the job ID in the KVS where it's already known (since it's one eventlog per job there).

Could we look for a place to add this to all events, and only for the RPC events stream?

One idea would be to enclose each eventlog entry in an outer object, just for the RPC stream, e.g.
```json
{
   "id":1234,
    "event":{
      "timestamp":42.24,
      "name":"start"
    }
}
```

That way the original event object can remain unchanged, and continue to be efficiently `json_incref()`ed for the RPC stream.

We also wouldn't need to update all the event definitions in RFC 21.

garlick reviewed (commented):
Couple initial comments that affect protocol - wanted to get those out and see what you think before proceeding.

chu11 commented:
rebased and re-pushed based on the comments from @garlick above.

chu11 commented:
had one builder fail with one of my new tests

> not ok 6 - job-manager: events with filters.deny works (multiple)

The output from the test is correct, so under the assumption travis was slow and the failure was on getting ENODATA in 5 seconds.  If this becomes common, I'll bump to 10s or bigger.

garlick reviewed (commented):
I'll have another look tomorrow - just wanted to get these few comments out before bed.

chu11 commented on src/modules/job-manager/event.c:
```diff
@@ -596,12 +715,202 @@ int event_job_post_pack (struct event *event,
+    json_t *allow = NULL;
+    json_t *deny = NULL;
+
+    if (flux_request_unpack (msg, NULL, "{s:o}",
+                             "filters", &filters) < 0)
```
oh duh, was "following" my prior code, didn't realize I could collapse this code :-)

chu11 commented:
re-pushed with fixes per @garlick comments above.  I didn't squash the fixes at the moment, LMK if you'd like me to squash to make review easier.

Also added one additional commit on top.  I realized there was a racy part in the `t2211-job-manager-events` test.

garlick commented:
Sure please squash. Thanks!

garlick commented on src/modules/job-manager/event.c:
```diff
@@ -596,12 +715,202 @@ int event_job_post_pack (struct event *event,
+            if (!strcmp (sender, tmpsender)) {
+                /* cannot remove from zlist while iterating, so we
+                 * store off entries to remove on another list */
+                if (!tmplist && !(tmplist = zlist_new ())) {
+                    errno = ENOMEM;
+                    flux_log_error (h, "%s: zlist_new", __FUNCTION__);
+                    goto error;
+                }
+                if (zlist_append (tmplist, el) < 0) {
+                    errno = ENOMEM;
+                    flux_log_error (h, "%s: zlist_append", __FUNCTION__);
+                    goto error;
+                }
```
Distinguishing zlist_append failure from the zlist_new failure in logs is not helpful, and `tmpsender` is leaked if either fails.  Maybe a helper function?
```c
static int create_zlist_and_append (zlist_t **lp, void *item)
{
    if (!*lp && !(*lp = zlist_new()))
        return -1
    if (zlist_append (*lp, el) < 0)
        return -1;
    return 0;
}
```
Then just call
```c
if (create_zlist_and_append (&tmplist, el) < 0) {
    flux_log (h, LOG_ERR, "job-manager.disconnect: failed to remove event listener: out of memory");
    zlist_destroy (&tmplist);
    goto error;
}
```

garlick commented on src/modules/job-manager/event.c:
```diff
@@ -200,6 +252,8 @@ void event_batch_destroy (struct event_batch *batch)
                                batch->annotations);
             json_decref (batch->annotations);
         }
+        if (batch->events_to_send)
+            send_events_to_listeners (batch->event);
         if (batch->responses) {
             flux_msg_t *msg;
             flux_t *h = batch->event->ctx->h;
```
I was initially slightly confused by this code, since `send_events_to_listeners()` is sandwiched between code that publishes events, and code that sends deferred responses.  It might be good to:

* rename `events_to_send` flag to something like `listener_response_available`
* rename  `send_events_to_listeners()` to something like `generate_listener_responses()`
* have that function call `event_batch_respond()` to create response messages and append to `batch->responses`, and then let the next block of code send the responses instead of repeating it

Also the comment block should be updated at the top of `event_batch_destroy()`.

chu11 replied:
ahhh, i did not know about the event batch responses

garlick commented on src/modules/job-manager/event.c:
```diff
@@ -596,12 +715,202 @@ int event_job_post_pack (struct event *event,
+    events_listener_destroy (el);
+}
+
+bool check_cancel (struct events_listener *el,
```
On the same note `add_entry_to_listener()` doesn't add anything.  Maybe `allow_deny_check()` or similar?

chu11 commented:
rebased & re-pushed with changes from @garlick's comments above

chu11 commented:
re-pushed, found a whitespace typo.  already squashed it.

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -443,7 +443,20 @@ static int eventlog_lookup_parse (struct info_ctx *ctx,
+        }
+        else if (!strcmp (name, "priority")) {
+            if (!context) {
+                flux_log_error (ctx->h, "%s: no priority context for %ju",
```
errno is not valid here so don't use `flux_log_error()` (nor in submit case).

garlick replied:
Probably should fix the submit case above too.

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -443,7 +443,20 @@ static int eventlog_lookup_parse (struct info_ctx *ctx,
+                goto nonfatal_error;
+            }
+
+            if (json_unpack (context, "{ s:i }",
+                             "priority", &job->priority) < 0) {
+                flux_log_error (ctx->h, "%s: priority context for %ju invalid",
+                                __FUNCTION__, (uintmax_t)job->id);
+                goto nonfatal_error;
+            }
```
`json_unpack()` doesn't set errno, so don't use `flux_log_error()`.

garlick replied:
Not to get too off track but the identical code right above it has the same problem. Fix here?

chu11 replied:
doh!  I could have swore i fixed these.  perhaps i rebased wrong.

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -174,6 +224,39 @@ struct job_state_ctx *job_state_create (flux_t *h)
+            flux_log_error (h, "error parsing events_reconnect_waittime");
+    }
+
+    if (!(w = flux_timer_watcher_create (r,
+                                         jsctx->events_reconnect_waittime,
+                                         0.,
+                                         setup_events_timer_cb,
+                                         jsctx))) {
+        flux_log_error (h, "flux_timer_watcher_create");
+        goto error;
+    }
```
I think you should just call `flux_reactor_stop_error()` if you ever get an error on the job manager event stream instead of trying to recover from it.  It seems unreasonable to expect modules to recover if services they use are unloaded.  If we decided to add this kind of recovery code across the board, we would expend a lot of effort that would be better applied toward formalizing the "used by" relationship between modules and preventing an "in use" module from being unloaded.

garlick replied:
In `job_events-continuation()`, only the first `flux_rpc_get_unpack()` handles an error by stopping the reactor, but every other error in that function would also be a malformed message, so shouldn't they all behave the same?

Also there is logging for each unique "protocol error" failure there.  I don't think that's useful for all these unlikely errors.

`job_events_continuation()` calls `job_events_priority()` which does some more parsing, but that function returns void.  Suggest you make that function return -1 on failure, and similarly, the logging of each detailed protocol failure probably isn't helpful.

chu11 replied:
ahhh correct.  I was "following" what I did before when checking for ENOSYS, but you are correct, the error should be handled everywhere.

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -214,6 +297,10 @@ void job_state_destroy (void *data)
+        if (jsctx->events)
+            flux_future_destroy (jsctx->events);
+        if (jsctx->events_reconnect)
+            flux_watcher_destroy (jsctx->events_reconnect);
```
Test for non-NULL is unnecessary for both of these destructors (and generally for all public flux API functions - it's a bug if we've missed any)

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -1673,6 +1760,111 @@ int job_state_init_from_kvs (struct info_ctx *ctx)
+    int priority;
+
+    if (!context) {
+        flux_log_error (jsctx->h, "%s: no priority context for priority event",
```
errno not set, don't call `flux_log_error()`

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -1673,6 +1760,111 @@ int job_state_init_from_kvs (struct info_ctx *ctx)
+    }
+
+    if (json_unpack (context, "{ s:i }", "priority", &priority) < 0) {
+        flux_log_error (jsctx->h, "%s: priority event context invalid",
```
errno not set by `json_unpack()`, don't call `flux_log_error()`

garlick replied:
Missed that one.

chu11 replied:
ugh, i went ahead and did a big audit of the whole file.  A bunch were due to cut & pastes from other areas.  I'll add as a separate commit.

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -1673,6 +1760,111 @@ int job_state_init_from_kvs (struct info_ctx *ctx)
+        json_t *context = NULL;
+
+        if (json_unpack (value, "{s:I s:o}", "id", &id, "entry", &entry) < 0) {
+            flux_log_error (jsctx->h, "%s: error parsing record",
```
errno is not set by `json_unpack()`, don't use `flux_log_error()`.

garlick commented on t/t2211-job-manager-events.t:
```diff
@@ -0,0 +1,181 @@
+test_under_flux 4
+
+RPC=${FLUX_BUILD_DIR}/t/request/rpc
+EVENT_STREAM=${FLUX_BUILD_DIR}/t/job-manager/event_stream
```
I think you could just use `t/request/rpc_stream` here instead of adding a new C program for these tests.  It's used by  `t2310-resource-module.t` and `t0025-broker-state-machine.t` to test test their streaming RPCs (for example).  Feel free to add options if you need them.

Or - a  python one liner that uses the Flux API is a very low-complexity way to do something one-off in a test like this. @grondo prodded me to use one in `t0001-basic.t` e.g.

```
test_expect_success 'flux-start with size 1 has no peers' '
        flux start ${ARGS} --size=1 \
                flux python -c "import flux; print(flux.Flux().rpc(\"overlay.lspeer\").get())" >idle.out &&
        test_must_fail grep idle idle.out
```

chu11 replied:
> I think you could just use t/request/rpc_stream here instead of adding a new C program for these tests.

I originally did, but wanted to cover the `job-manager.events-cancel` case, which `rpc_stream` doesn't.

Let me actually update `event_stream` to output each event separately on a new line.  It'll be necessary when I convert `job-info` to use annotations.  (The `alloc` eventlog entry includes an annotation in the context, so using `grep` will no longer be useful in testing).

garlick reviewed (changes requested):
job manager changes are looking good!  Thanks for that iteration.
Now some review on the job info side.

chu11 commented:
rebased & re-pushed based on comments above.  Sorry for the lazy cut & paste mistakes.

- with the removal of the reconnect code in `job-info`, a nice chunk of code is gone now.  e.g.

  - The "setup_event_listeners()` function isn't needed anymore for example (is only used in 1 place now).

  - a nice chunk of testing code is gone too.  The `job-manager-helper.sh` script is no longer needed (the "helper" functions are now only used in 1 file now).

  - I only call `flux_reactor_stop_error()` on `!= EPROTO`, feeling that EPROTO is an acceptable "survivable" error.

- as a result of the above, we need to load the `job-info` module before the `job-manager` now, otherwise `job-info` will get `ENOSYS` a bunch.  Adjusted `rc` scripts accordingly.  A nice chunk of tests had to be adjusted too (i.e. on reload of `job-manager`, also have to reload `job-info`).

- the `event_stream` testing tool now outputs each event on a separate line and the tests in `t2211-job-manager-events` now uses `jq` to parse each line and make sure the event `.name` matches what is expected instead of just using grep.

  this change is unnecessary for this PR, but as I began annotations work in a different branch, I realized it would be necessary since multiple annotation "events" can occur and I could no longer just "grep" like I had before.  So I just merged those fixes in here.

- added a cleanup commit that removes some unnecessary NULL checks before calling `zlistx_destroy()` or `zhashx_destroy()`.

- went ahead and squashed, hope that's ok

garlick commented:
Thanks!  I'm packing up this morning to travel and am not sure I will have time to get through a review before I hit the road, so couple quick comments on your comments:

> I only call flux_reactor_stop_error() on != EPROTO, feeling that EPROTO is an acceptable "survivable" error.

Why is EPROTO survivable?  Wouldn't that indicate a coding error?

> as a result of the above, we need to load the job-info module before the job-manager now, otherwise job-info will get ENOSYS a bunch. Adjusted rc scripts accordingly.

Wait, why?  job-info is making the RPC to job-manager, so job-manager should be loaded first.  What am I missing?

chu11 commented:
just noticed a whitespace tabbing blunder during my cleanup, squashed & re-pushed.

chu11 commented:
>>     as a result of the above, we need to load the job-info module before the job-manager now, otherwise job-info will get ENOSYS a bunch. Adjusted rc scripts accordingly.

>Wait, why? job-info is making the RPC to job-manager, so job-manager should be loaded first. What am I missing?

OOPS OOPS OOPS!  Talk about typos ... load job-info after job-manager.  Remove job-info before job-manager .... :P

> Why is EPROTO survivable? Wouldn't that indicate a coding error?

Yeah it would.  I was just trying to think of possible errnos that would be "ok", and EPROTO was the only one that came to mind.  I'll adjust for all errnos.

garlick commented:
Ah, I assumed job-manager was already loaded first, so I didn't catch that that might be a typo. (For a bit I thought I had seriously misinterpreted some code!)

I'll need to pick this up tomorrow AM as the rest of my day is  pretty much spoken for.  Feel free to tweak in the mean time - i don't have a review in progress or any state that is going to get messed up by a force push.

chu11 commented:
re-pushed with just some stupid fixes.  realized some functions could be `static`.  Removed some dead code in the tests.

chu11 commented:
hmmm `flux-sched` build failed with `FAIL: t5000-valgrind.t 1 - valgrind reports no new errors on 2 broker run`.  Unlikely my changes would affect things.  Re-built locally and `t5000-valgrind` passed  multiple times.

chu11 commented:
re-pushed, found a stupid typo

chu11 commented:
re-pushed, minor cleanups on the test titles

chu11 commented:
re-pushed, realized i could make some of the testing helper functions faster

chu11 commented:
re-pushed, just another tiny test tweak

garlick reviewed (approved):
Thanks, looking goood!

I mentioned some error handling cleanup in the two unresolved comments, but this is very close.  I'll go ahead and mark approved.

There should be some further opportunities to simplify job-info by using this new job manager RPC.  It would be good to identify those in an issue and then we can decide if it makes sense to jump on those right away.

chu11 commented:
@garlick .  Cool, thanks!

> There should be some further opportunities to simplify job-info by using this new job manager RPC

yup, I already got a list of follow ups

- send annotations w/ `job-manager-events`, thus removing `job-annotations event`.
- add event caching (i'm calling it caching instead of circular buffer in #3209)
- the big one ... consider refactoring to use only job-manager.events and not `job-state`.  I have to think about this more if its a wise idea.

garlick commented:
> the big one ... consider refactoring to use only job-manager.events and not job-state. I have to think about this more if its a wise idea.

I hope so, otherwise this mechanism wasn't the good idea I thought it was...

chu11 commented:
re-pushed with all of the fixes talked about above, squashed

chu11 commented:
> I hope so, otherwise this mechanism wasn't the good idea I thought it was...

It minimally solves the "update job-info with info about changes in the job-manager" which is the important thing.  I think it'll work out, is just 1 or 2 steps ahead and hard to see the end at the moment.

garlick commented:
If you're ready I'm fine with MWP!

chu11 commented:
@garlick thanks
